src/widget/wbasewidget.h

```diff
@@ -95,7 +126,13 @@ class WBaseWidget {
 
   private:
     QWidget* m_pWidget;
+
     QString m_baseTooltip;
+    QString m_shortcutTooltip;
+    QString m_baseTooltipOptShortcuts;
+    QList<std::pair<ConfigKey, QString>> m_shortcutKeys;
```

Comment:
For knobs and sliders there'll be 5 controls (control, _up, _down, _up_small, _down_small).
For other widgets it's 1-3 (left- and right-click, maybe a display conection).

Reply:
thanks. makes sense. This is not duplicated between widgets then?

Reply:
No. It's set by LegacySkinParser and fetched by KeyboardEventFilter which looks up the keys per control, composes the tooltip and pushes that to the widget.

Reply:
Btw this QList with a pair feels a bit clunky, so I switched to a QMap<ConfigKey, QString> (sorted by key) with convenient cons_key_value_iterator.
This also has the benefit that shortcuts are now always sorted by control/sub-control's ConfigKey¹ (same group, so sorted by item) instead of order they've been added (QList) or sorted by the command's _tr_ string (std::map<QString, ConfigKey>.
The slight performance regression shouldn't matter much as we add 4 items max.

¹WPushButton: _activate, _toggle
WKnob/WSlider: _down, _down_small, _up, _up_small

Reply:
Oh, just noticed that we don't add the other ControlPotmeter controls (_set_minus_one, _set_default etc.)
When I add those for sake of completeness, it'd be 10 sub-controls per knob/slider, and we have a lot of knobs (fx parameters, mixer, maybe 64 samplers, ...)

Will revert to QList<std::pair<QString, ConfigKey)

Reply:
and we're also missing the HotcueButton controls (activate, clear + set, goto, .., gotoloop) which would add another 12 controls per button : \

Note that both these extra ControlPotmeter and WHotcueButton controls are not used in built-in mappings, except `rate_perm` and `rate_temp` of deck 1/2 (which are mapped to dedicated buttons in official skins anyway).

Not sure how to proceed, ie. idk whether there's a noteworthy performance impact (when toggling) keyboard shortcuts) if we add all those sub-controls which are likely not used.
What do yo think?

Reply:
I'm primarily concerned about the memory overhead from many large QLists in many widgets. But I can't make a good judgement call here either.

Reply:
Right, but I was considering memory overhead (likely also not easily measurable because we're wasting so much already in other places), not runtime performance. Moreover 80ms is quite a lot...

Reply:
> Moreover 80ms is quite a lot...

ah okay, without the qdebug lines it's down to ~4ms for ~1600 widgets (LateNight).

I'll push the latest version soonish.

Reply:
...and now (without the debug stuff) it's obvious that the extra controls do have a noticeable impact:
with: 4-5ms
without: ~2.5ms